src/libraries/System.Net.Primitives/src/System/Net/IPNetwork.cs

```diff
@@ -0,0 +1,344 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Net.Sockets;
+
+namespace System.Net
+{
+    /// <summary>
+    /// Provides an Internet Protocol (IP) subnet/range in CIDR notation.
+    /// </summary>
+    public readonly struct IPNetwork : IEquatable<IPNetwork>, ISpanFormattable, ISpanParsable<IPNetwork>
+    {
+        public IPAddress BaseAddress { get; private init; }
+        public int PrefixLength { get; private init; }
+
+        /*
+         We can also add two more easy to implement properties:
+         * AddressCount - a lazily calculated property showing the amount of potential address this instance "Contains". 2^(128_OR_32 - PrefixLength)
+            * 2^128 is going to fit into BigInteger, which causes a concern for memory allocation, since for actually big integers BigInteger seems to allocate a byte array.
+            * We can just have a Nullable<BigInteger> backing field for the lazy property. But we should be fine with the increased byte size for IPNetwork type itself caused by that.
+         * Since BaseAddress represents the first address in the range, does it make sense to have a property that represents the last address?
+         */
+
+        private const char addressAndPrefixLengthSeparator = '/';
+        private const int bitsPerByte = 8;
+        private const string
+            baseAddressConstructorParamName = "baseAddress",
+            prefixLengthConstructorParamName = "prefixLength";
+        public IPNetwork(IPAddress baseAddress, int prefixLength)
+        {
+            BaseAddress = baseAddress;
+            PrefixLength = prefixLength;
+
+            var validationError = Validate();
+            if (validationError.HasValue)
+            {
+                throw validationError.Value.ConstructorExceptionFactoryMethod();
+            }
+        }
+
+        public bool Contains(IPAddress address)
+        {
+            ArgumentNullException.ThrowIfNull(address);
+
+            if (address.AddressFamily != BaseAddress.AddressFamily)
+            {
+                return false;
+            }
+
+            // TODO: this can be made much easier and potentially more performant for IPv4 if IPAddress.PrivateAddress is made internal (currently private)
```

Comment:
@karelz Can we make [IPAddress.PrivateAddress](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Primitives/src/System/Net/IPAddress.cs#L70) internal for this?

Reply:
You can access `IPAddress.Address` instead, as that's already public (you can suppress that obsoletion warning after checking that the IP is indeed IPv4)

Reply:
@MihaZupan seems like IPAddress.Address is constructed in a way that the left high order byte in IPAddress.Address is the last segment of and an IPv4 address. So we have a right-to-left flow when it comes to the bytes in IPAddress.Address, but the bits in individual bytes of IPAddress.Address are left-to-right (which messes with our ability to just go through bits right-to-left).

```
var address = IPAddress.Parse("255.255.255.64");
		
Console.WriteLine(Convert.ToString(address.Address, 2).PadLeft(32, '0')); // 01000000111111111111111111111111
Span<byte> span = stackalloc byte[4];
address.TryWriteBytes(span, out _);
foreach(var segment in span)
{
	Console.Write(Convert.ToString(segment, 2).PadLeft(8, '0'));
}
Console.WriteLine(); // 11111111111111111111111101000000, this is the order we need
```

So it seems like we will have the same implementations for both IPv4 and IPv6